### ISSUE TITLE
[Proposal]: struct members assign operator `<=`

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#513]
I suggest to solve [this issue]( https://github.com/dotnet/vblang/issues/511) and [this](https://github.com/dotnet/vblang/issues/487) by generally solving the whole issue for structures, by adding the `<=` operator:
```VB.NET
Dim l = new List(Of (X As Integer, Y As Integer)
l.Add((1,1))
l(0).X <= 2 
```

Which is lowered to:
```VB.NET
var temp = l(0)
temp.x = 2
 l(0) = temp
```

**Comment by tverweij** [comment]
Whats wrong with:

Dim l = new List(Of (X As Integer, Y As Integer)
l.Add((1,1))
l(0) = (2, l(0.y))

**Comment by VBAndCs** [comment]
Less readable.
Besides, it will be insane to do the same with a 10-item tuple.
Furthermore, it is more complex with properties that return a structure that needs to re-initialize all members to have this inline style, which will force you to use a temp var to do a simple assignment in 3 lines.
Same with properties that return arrays:
```
Dim L = TextBox.Lines(1) 
L(1) = "Hello"
TextBox.Lines = L 
```

Why not:
`TextBox.Lines(1) <= "Hello"`

**Comment by tverweij** [comment]
I will suggest this one. No promises.

**Comment by WolvenRA** [comment]
> Why not:
> `TextBox.Lines(1) <= "Hello"`

Why not:
`TextBox.Lines(1) = "Hello"`

**Comment by VBAndCs** [comment]
Using = will set the array, but will not invoke the lines peoperty setter, and .Net gives syntax error to prevent this confusion. I don,t want to change how = works.

**Comment by gilfusion** [comment]
Another option:

You said that ```l(0).x <= 2``` is equivalent to:
```vb.net
Dim temp = l(0)
temp.x = 2
l(0) = temp
```
Those first two lines could be simplified by adding a ```With``` expression (generalizing from object initializers--I'm sure there's an issue for it around here somewhere).
```vb.net
Dim temp = l(0) With {.x = 2}
l(0) = temp
```
or 
```vb.net
l(0) = l(0) With {.x = 2}
```
And, like how ```i = i + 1``` gets simplified to ```i += 1```, you could even create a simplification here to
```vb.net
l(0) With= {.x = 2}
```

**Comment by rskar-git** [comment]
This proposal comes from a point of major confusion: Members of a `Structure` (including `ValueTuple`) are already assignable; what some folks are actually experiencing is an l-value versus an r-value situation. An l-value generally means a variable or field, and it is meant to represent a definitive place (in memory or CPU register) in which to store an instance of whatever the l-value is meant to store. 

Properties, by design, work with copies, not originals, and hence will never make available the actual location of any of their elements. Hence, there is no way for a value type coming from a property to then have any assignable members.

What may muddle things for some folks would be over the true nature of reference types. With a statement like this:

    Dim o As ListBoxItem = New ComboBoxItem()
	
The `New ComboBoxItem()` is the expression that creates an instance of ComboBoxItem, and then returns its reference (which itself is a peculiar kind of value type). So a reference type has this dichotomy: (1) An instance, somewhere in memory; and (2) a reference, the location information on that instance. Variable `o` does NOT contain the instance; it contains the reference. Saying `o As ListBoxItem` means whenever we mention `o`, we mean to query the reference it contains for the `ListBoxItem` aspect of the instance. Since `ComboBoxItem` inherits from `ListBoxItem`, that works.

	' One needs location information to do an assignment
	Dim l = New List(Of (X As Integer, Y As Integer, Z As Text.StringBuilder))
	l.Add((1, 1, New Text.StringBuilder("abcdefg")))
	l(0).Z.Length = 3 ' This works, because Z is reference, hence location information.
    l(0).Z = New Text.StringBuilder() ' Nope, not allowed, Z is member of value-type.
	l(0).X = 2 ' This does NOT work, because value-types are devoid of location information.

Note statements like these that work today:

    Dim v As (X As Integer, Y As Integer)
    v = (1, 1)
    v.X = 123 ' This compiles, v and v.X are proper l-values

    Dim v(9) As (X As Integer, Y As Integer)
    v(0) = (1, 1)
    v(0).X = 123 ' This DOES compile, and works, because v(0) is a proper l-value

A `List` or `IList` or any `Class` with an indexer (`Default Property`) may seem the same as `Array`, but they are not. `Array` is special because it is defined in very specific terms of a base location and integer-based indexing; by design, the compiler will always have reliable access to the location of any element in an array.

The only way to make this proposal work properly is to put deeper dependencies into the compiler for `IList` and `IList(Of T)`. In other words, the compiler will need major tweaks to somehow treat `IList` as if a proper `Array`, and to somehow do that without reliable location information at hand at compile time. That's a super cringy feature to make work.

That lowered code suggested within this proposal will bring about headaches of not only behind-the-scenes temporary variables getting littered about, but also overdone copy operations. E.g. imagine a situation where someone would like to modify a few members within a several-member tuple?

	// Imagine this lowered code for a (X,Y,Z,U,V) tuple
	// l(0).X <= 2;
	{ var temp = l(0); temp.X = 2; l(0) = temp;}
	// l(0).Y <= 3;
	{ var temp = l(0); temp.Y = 3; l(0) = temp;}
	
In the grand scheme of things, using a temporary variable for this situation already works well. 

	// And now as sensibly done by a human being
	// l(0).X <= 2; l(0).Y <= 3;
	{ var temp = l(0); temp.X = 2; temp.Y = 3; l(0) = temp;}

**Comment by VBAndCs** [comment]
@rskar-git  
An optimizer can solve this, and we can help it like this:
```
With l(0)
    .X <= 2
    .Y <= 3
End With
``

To produce:
```
Dim temp = l(0)
temp.X = 2
temp.Y = 3
 l(0) = temp
```

This a one place where I can start using With blocks and feel it is helpful.

**Comment by rskar-git** [comment]
>...we can help it like this `With l(0)`...

(@gilfusion suggested a `With` expression, but I find that to be problematic with reference types.)

Actually if powers of the `With` statement could be extended... In the context of a `With` statement, perhaps something that would be suitable as a `ByRef` parameter in a `Sub` call might be workable. There would be no need for a new operator, equals sign would be fine; nor any need to hope for some magical optimizer!

So this, which is currently impossible to do:

    With l(0) ' l(0) is a property/indexer returning unassignable value-type
        .X = 2
        .Y = 3
    End With
    
might someday be possible via lowered code which may resemble this:

    Call Sub(ByRef WithVar As (X As Integer, Y As Integer))
             With WithVar
                 .X = 2
                 .Y = 3
             End With
         End Sub(l(0))

So basically, `With` needs to detect when a write-able value-type property with parameter(s) is present to make use of this alternative implementation. If one wants to force current behavior, then the expression needs to be inside parenthesis (similar to how `ByRef` can be undone in a `Sub` or `Function` call today):

    With (l(0)) ' Force value only mode; assignment now error again
        .X = 2  
        .Y = 3
    End With

The only thing to optimize on then is on whether any assignments occur - if there are none, do not use the new alternative implementation.
